apps/sms/services/js/moz_mobile_message/moz_mobile_message_shim.js

```diff
@@ -0,0 +1,231 @@
+/* -*- Mode: js; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- /
+/* vim: set shiftwidth=2 tabstop=2 autoindent cindent expandtab: */
+
+
+/* global bridge */
+
+'use strict';
+(function() {
+
+/**
+ * Name of the service for mozMobileMessage API shim
+ * @type {string}
+ */
+
+const SERVICE_NAME = 'mozMobileMessageShim';
+
+/**
+ * Array of the event name that is corresponding to mozMobileMessage events.
+ * @type {Array}
+ */
+
+const EVENTS = ['received', 'sending', 'sent', 'failed', 'deleted',
+                'readsuccess', 'deliverysuccess'];
+
+/**
+ * Array of method that need to be exposed for API shim.
+ * @type {Array}
+ */
+
+const METHODS = ['getMessage', 'retrieveMMS', 'send', 'sendMMS',
+                 'delete', 'markMessageRead', 'getSegmentInfoForText'];
+
+/**
+ * Array of stream that need to return data(messages/threads) in chunk.
+ * @type {Array}
+ */
+
+const STREAMS = ['getThreads', 'getMessages'];
+
+var mozMobileMessage = navigator.mozMobileMessage;
+var service = bridge.service(SERVICE_NAME);
+
+var MozMobileMessageShim = {
+  init: function ms_init() {
+    EVENTS.forEach((event) => {
+      mozMobileMessage.addEventListener(
+        event,
+        this[Utils.camelCase(`on-${event}`)]
+      );
+    });
+
+    METHODS.forEach((shimMethod) => {
+      service.method(shimMethod, this[shimMethod]);
+    });
+
+    STREAMS.forEach((shimStream) => {
+      service.stream(shimStream, this[shimStream]);
+    });
+  },
+
+  /* Events */
+
+  onSending: function ms_onSending(e) {
+    service.broadcast('message-sending', { message: e.message });
+  },
+
+  onFailed: function ms_onFailed(e) {
+    service.broadcast('message-failed-to-send', { message: e.message });
+  },
+
+  onDeliverysuccess: function ms_onDeliverysuccess(e) {
+    service.broadcast('message-delivered', { message: e.message });
+  },
+
+  onReadsuccess: function ms_onReadsuccess(e) {
+    service.broadcast('message-read', { message: e.message });
+  },
+
+  onSent: function ms_onSent(e) {
+    service.broadcast('message-sent', { message: e.message });
+  },
+
+  onReceived: function ms_onReceived(e) {
+    var message = e.message;
+
+    if (message.messageClass && message.messageClass === 'class-0') {
```

Comment:
I'm not sure either -- this looks "business" already, I think this check would be in MessagingService instead